### PR TITLE
[Merged by Bors] - ATX Builder overhaul

### LINKED ISSUE #1061: Self ATXs should be stored in DB when created (and not when received on Gossip)
# Overview / Motivation
Currently, we store all ATXs (self and remote) once they appear on Gossip. If we crash between the time we sent a self ATX and the time we receive and store the ATX we might get several races - we might re-create the same ATX once we restarted (and risk being black flagged) or we might try to start working on the next ATX with the wrong ATX as the previous ATX.

# The Task
Store state of the self-ATX in DB before committing (sending) to the ATX.

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who have'nt introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### LINKED ISSUE #1204: ATX builder shouldn't store previous atx
Currently, the atx builder stores the most recent atx that was generated in order to use it as the previous atx when coming to generate the next atx. This causes bugs where the atx builder uses the cached atx although (due to other bugs) the atx was never introduced in any block.
This cache is useless since we anyway need to validate that the atx is in the db before using it.

### LINKED ISSUE #1208: Miner wait a full layer after publishing an atx, before he starts working on the next atx
Should be right after publication -- unless we want to make sure that our atx is in a block, before we start working on the following atx

### LINKED ISSUE #1494: Don't discard NIPoST who's publication epoch is below the current
When building an ATX, we discard a NIPoST if its publication epoch is below the current epoch. This should be changed to allow the usage of a NIPoST also in the next epoch. 
So for example, if the publication epoch is 2 to earn eligibility to produce blocks in epoch 3, it would be good to publish it in epoch 3 as well. Only if we are in epoch 4 the NIPoST should be discarded.

### LINKED ISSUE #1550: Nipst builder can get stuck on waiting for poet proof
# Overview / Motivation
NIPST builder can get stuck on waiting for results from poet for too long. 
we need to think about when we stop waiting for membership proofs and restart the whole process. 

# The Task
implement a timeout on waiting for poet membership proofs .
# Implementation Notes\
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who haven't introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### LINKED ISSUE #1551: recover from invalid ATX
# Overview / Motivation
Currently, an ATX is not tested for validity when publishing into network. 
it is then refrenced by the next ATX created, thus invalidating the next ATX as well.

# The Task
Validate the ATX when creation is complete, if ATX is invalid (investigate) and re run process.

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who haven't introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### LINKED ISSUE #1552: ATX should not be sent after 2 epoch has passed.
# Overview / Motivation
ATX that was created 2 epochs in the past shouldn't be brodcasted to the network. 

# The Task
Discard atx if it's publication index is 2 epochs in the past.

# Implementation Notes
TODO: Add links to relevant resources, specs, related issues, etc...

# Contribution Guidelines
Important: Issue assignment to developers will be by the order of their application and proficiency level according to the tasks complexity. We will not assign tasks to developers who haven't introduced themselves on our Gitter [dev channel](https://gitter.im/spacemesh-os/Lobby)

1. Introduce yourself on go-spacemesh [dev chat channel](https://gitter.im/spacemesh-os/Lobby) - ask our team any question you may have about this task
2. Fork branch `develop` to your own repo and work in your repo
3. You must document all methods, enums and types with [godoc comments](https://blog.golang.org/godoc-documenting-go-code)
4. You must write go unit tests for all types and methods when submitting a component, and integration tests if you submit a feature
5. When ready for code review, submit a PR from your repo back to branch `develop`
6. Attach relevant issue to PR

### PR DESCRIPTION
**Pull request by noamnelke**
## Motivation
Closes #1550 
Closes #1551 
Closes #1552 
Closes #1204 
Closes #1208 
Closes #1061 
Closes #1494 
Related to #1596, but only partially handles it

## Changes
The node will not start working on an ATX until the previous ATX was included in a block and added to the DB.
Instead of trying to publish an ATX at the start of each layer, ATX creation happens in a loop. When certain required conditions don't apply, execution waits until they do.

The flow is now as follows:
- If a NIPST challenge is stored, but is stale -- clear it.
- If a NIPST challenge is not stored -- create one.
  - Ensure the node is synchronized, or wait if it isn't.
  - Get a positioning ATX.
  - Get the miner's previous ATX.
- Ensure the publication epoch has started, so we can include a full view of the positioning epoch.
- Ensure the node is synchronized.
- Get the view.
- Calculate the active set size from the view.
- Compose, sign and publish the ATX.
- Wait until the ATX is seen in a block.
  - If the ATX's target epoch has passed, wait until the node is synchronized and if it still hasn't been seen in a block consider it an error and start over.
  - If this happens, since the previous ATX is always taken from the database, the node will create a new ATX to replace the one that failed.

## Test Plan
Unit tests were added. All existing tests (unit + automation) pass.

### CODE REVIEW

**Comment by noamnelke** [comment]
bors try

File: miner/atx_pool.go [hunk 1]
```diff
@@ -39,6 +40,9 @@ func (mem *AtxMemPool) GetAllItems() []*types.ActivationTx {
 func (mem *AtxMemPool) Put(atx *types.ActivationTx) {
 	mem.mu.Lock()
 	mem.atxMap[atx.Id()] = atx
+	if mem.listener != nil {
+		go mem.listener(atx.ActivationTxHeader)
```
Comment (antonlerner) [inline miner/atx_pool.go]: this is called for every atx we get in the pool, not only the one we are waiting for

Comment (antonlerner) [reply miner/atx_pool.go]: add a test for this case as well

File: activation/activation.go [hunk 2]
```diff
@@ -467,36 +478,57 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
...
-	// cleanup state
+func (b *Builder) cleanupState() {
 	b.nipst = nil
 	b.challenge = nil
 	b.posLayerID = 0
-
-	err = b.net.Broadcast(AtxProtocol, buf)
-	if err != nil {
-		return err
+	if err := b.discardChallenge(); err != nil {
```
Comment (antonlerner) [inline activation/activation.go]: what about unergistering the hook if we haven't received it?

File: activation/activation.go [hunk 3]
```diff
@@ -467,36 +478,57 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
 	b.log.With().Info("active ids seen for epoch", log.Uint64("pos_atx_epoch", uint64(posEpoch)),
 		log.Uint32("view_cnt", activeSetSize))
 
-	signedAtx, err := b.SignAtx(atx)
-	if err != nil {
+	atxReceived := make(chan struct{})
+	b.atxPool.SetListener(func(atxHeader *types.ActivationTxHeader) {
```
Comment (antonlerner) [inline activation/activation.go]: a more go approach is to pass a channel and not a delegate

File: activation/activation.go [hunk 4]
```diff
@@ -467,36 +478,57 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
...
+		log.EpochId(uint64(atx.PubLayerIdx.GetEpoch(b.layersPerEpoch))),
+		log.Uint32("active_set", atx.ActiveSetSize),
+		log.String("miner", b.nodeId.Key[:5]),
+		log.Int("view", len(atx.View)),
+	}
+	select {
+	case <-atxReceived:
+		b.log.Event().Info("atx published!", fields...)
+	case <-time.After(b.broadcastTimeout):
+		b.log.With().Error("broadcast timeout when publishing ATX", fields...)
```
Comment (antonlerner) [inline activation/activation.go]: should we discard the current challenge?

File: miner/atx_pool.go (hunk 1 above)
Comment (antonlerner) [reply miner/atx_pool.go]: put is also called without any checks, so I think that potentially you can get `mem.listener` to be called twice and perhaps even panic with double close on the channel, WDYT?

Comment (noamnelke) [reply miner/atx_pool.go]: Yes. It's the listener's job to only act on the relevant ATX.

cc: @gavraz

File: activation/activation.go (hunk 4 above)
Comment (noamnelke) [reply activation/activation.go]: I don't think. Most likely the failure was due to the previous ATX not being in the database yet (waiting for miners to include it and pbase to advance). By leaving the state we'll retry every layer.

Once we fix #1552, there will also be a termination condition to the retries.

File: activation/activation.go (hunk 3 above)
Comment (noamnelke) [reply activation/activation.go]: Not sure exactly what you mean.

File: activation/activation.go (hunk 2 above)
Comment (noamnelke) [reply activation/activation.go]: done.

bd26b7f

File: activation/activation.go (hunk 4 above)
Comment (antonlerner) [reply activation/activation.go]: if prevAtx was never included in the mesh this will never terminate until an epoch will pass

Comment (antonlerner) [reply activation/activation.go]: if prevAtx was never published and will never be on the mesh, an ATX will not be built in this epoch and only after state is discarded a new atx will be published. do you think we should handle this?

File: miner/atx_pool.go (hunk 1 above)
Comment (noamnelke) [reply miner/atx_pool.go]: Only now noticed your last comment.

> put is also called without any checks

Not true, `Put` locks the mempool, see line 41.

File: activation/activation.go (hunk 4 above)
Comment (noamnelke) [reply activation/activation.go]: You're right, we won't publish an ATX in the current epoch, but how can we tell that the previous ATX will **never** be published?

I think that this can only be fixed when we support free-standing ATXs. Until then there's always a risk of blacklisting yourself if you discard state.

@barakshani thoughts?

Comment (barakshani) [reply activation/activation.go]: 1. no need for pbase to advance - a block (and its content) is in the database once received (and syn. valid). Anyway, with free-standing that should change.
2. why a new atx will not be built/published if the previous is not on the mesh? is it an outcome of this PR's new code?

File: activation/activation.go [hunk 5]
```diff
@@ -176,9 +185,11 @@ func (b *Builder) loop() {
 				err := b.PublishActivationTx(epoch)
 				if err != nil {
 					if err == tooSoonErr {
```
Comment (barakshani) [inline activation/activation.go]: While this error is desired for our tests, it is not inline with the protocol.
There should be no reason to prevent someone from publishing an _early_ ATX, if they want to do so.
I think we should have a comment in the code that we should remove this condition once ready (once poet running time is calibrated with epoch length, and probably weights/ticks as well)

File: activation/activation.go (hunk 4 above)
Comment (barakshani) [reply activation/activation.go]: ok, so you meant this timeout condition. with this new code we never clear b.nipst, hence we will not try to build a new atx.
What would actually happen then? At the beginning of a layer, we enter the function and go to line 440. Would we try to re-broadcast the atx? This should be the desirable behaviour, along with issue #1552.
In my opinion, it is not super critical for testnet, and if the miner has not received the atx until the timeout, we can assume it is lost forever, and start building a new one, with _previous atx_ the last one actually published.

Comment (noamnelke) [reply activation/activation.go]: Cool, so this is what happens with this code.

Each layer we'll try to broadcast, and if it fails, just try again next layer. The only reason it should fail is due to the previous ATX not being in the ATXDB yet, which only currently happens when pbase passes a layer and the block is contextually valid.

Currently, this will happen forever. Once we implement #1552 we'll only try until the ATX is no longer relevant and then assume it's lost forever.

Once free-standing ATXs are implemented (#1059) this should really never happen since ATXs will be in the DB immediately, and we'll validate that an ATX hits the DB instead of the mempool, so the previous one will definitely be there already.

Comment (noamnelke) [reply activation/activation.go]: Cool, so this is what happens with this code.

Each layer we'll try to broadcast, and if it fails, just try again next layer. The only reason it should fail is due to the previous ATX not being in the ATXDB yet, which only currently happens when pbase passes a layer and the block is contextually valid.

Currently, this will happen forever. Once we implement #1552 we'll only try until the ATX is no longer relevant and then assume it's lost forever.

Once free-standing ATXs are implemented (#1059) this should really never happen since ATXs will be in the DB immediately, and we'll validate that an ATX hits the DB instead of the mempool, so the previous one will definitely be there already.

**Review by antonlerner — APPROVED** [review]

**Comment by barakshani** [comment]
As stated in https://github.com/spacemeshos/go-spacemesh/issues/1551, please consider https://github.com/spacemeshos/go-spacemesh/issues/1061 and https://github.com/spacemeshos/go-spacemesh/issues/1204 in the design of the solution.
Ideally, this PR should close all these issues.
@noamnelke @antonlerner

File: activation/activation_test.go [hunk 6]
```diff
@@ -334,23 +335,37 @@ func TestBuilder_PublishActivationTx_FaultyNet(t *testing.T) {
...
+	r.True(builtNipst)
+
+	// the next time this runs (an epoch + a layer later), the nipst is NOT re-built
+	published, builtNipst, err = publishAtx(b, postGenesisEpochLayer+layersPerEpoch+2, postGenesisEpoch+1, layersPerEpoch)
+	r.EqualError(err, "broadcast timeout")
+	r.False(published)
+	r.False(builtNipst)
+
+	// the next time this runs (an epoch + a layer later), the nipst is NOT re-built
+	published, builtNipst, err = publishAtx(b, postGenesisEpochLayer+(2*layersPerEpoch)+1, postGenesisEpoch+2, layersPerEpoch)
```
Comment (antonlerner) [inline activation/activation_test.go]: check that the nipst is discarded

Comment (noamnelke) [reply activation/activation_test.go]: Good catch - added a test for this.

File: activation/activation.go [hunk 7]
```diff
@@ -176,9 +180,11 @@ func (b *Builder) loop() {
...
-						b.log.Warning("cannot create atx in epoch %v: %v", epoch, err)
+						b.log.With().Warning("failed to publish ATX",
```
Comment (antonlerner) [inline activation/activation.go]: you can drop the if

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

**Comment by noamnelke** [comment]
bors try

File: activation/atxdb.go [hunk 8]
```diff
@@ -446,6 +472,13 @@ func (db *ActivationDb) storeAtxUnlocked(atx *types.ActivationTx) error {
 	if err != nil {
 		return err
 	}
+
+	// notify subscribers
+	if ch, found := db.atxChannels[atx.Id()]; found {
+		close(ch)
+		delete(db.atxChannels, atx.Id())
```
Comment (antonlerner) [inline activation/atxdb.go:502]: you delete from map with no lock

File: activation/activation.go [hunk 9]
```diff
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
...
-						b.log.Error("cannot create atx in epoch %v: %v", epoch, err)
-					}
-					events.Publish(events.AtxCreated{false, "", uint64(epoch)})
-				}
-
-				b.finished <- layer
-			}()
-		case <-b.finished:
-			b.working = false
+		default:
```
Comment (antonlerner) [inline activation/activation.go:180]: why leave the default?

File: activation/atxdb.go (hunk 8 above)
Comment (noamnelke) [reply activation/atxdb.go:502]: This is inside the method `storeAtxUnlocked` which is only called under lock

File: activation/activation.go (hunk 9 above)
Comment (noamnelke) [reply activation/activation.go:180]: If I remove the `default:` it will block on the `stop` channel. By having a `default:` I check if `stop` has been closed, `return` if it did, and otherwise continue execution.

File: activation/activation.go [hunk 10]
```diff
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
...
-			if err != nil {
-				// TODO: handle inconsistent state
-				b.log.Panic("prevAtx (id: %v) not found in DB -- inconsistent state", prevAtxId.ShortString())
+		if err := b.PublishActivationTx(); err != nil {
+			if _, stopRequested := err.(StopRequestedError); stopRequested {
+				return
 			}
+			b.log.With().Error("failed to publish ATX", log.Err(err))
+			events.Publish(events.AtxCreated{false, "", uint64(b.currentEpoch())})
+			<-b.layerClock.AwaitLayer(b.layerClock.GetCurrentLayer() + 1)
```
Comment (antonlerner) [inline activation/activation.go:188]: why do we need to wait an entire layer to resend?

File: activation/activation.go [hunk 11]
```diff
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
...
+	} else {
+		challenge.PrevATXId = prevAtx.Id()
+		challenge.Sequence = prevAtx.Sequence + 1
 	}
-
-	err = b.storeChallenge(b.challenge)
-	if err != nil {
-		log.Error("challenge cannot be stored: %s", err)
+	b.challenge = challenge
+	if err := b.storeChallenge(b.challenge); err != nil {
```
Comment (antonlerner) [inline activation/activation.go:214]: shouldn't you return this error?

File: activation/activation.go [hunk 12]
```diff
@@ -383,60 +324,57 @@ func (b *Builder) loadChallenge() error {
...
-}
-
-// PublishActivationTx attempts to publish an atx for the given epoch, it returns an error if an atx cannot be created
-// publish atx may not produce an atx each time it is called, that is expected behaviour as well.
-func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
-	if b.nipst != nil {
-		b.log.With().Info("re-entering atx creation in epoch %v", log.EpochId(uint64(epoch)))
+// PublishActivationTx attempts to publish an atx, it returns an error if an atx cannot be created.
+func (b *Builder) PublishActivationTx() error {
+	b.refreshChallenge()
```
Comment (antonlerner) [inline activation/activation.go]: why is it called refreshChallenge?

File: activation/activation.go [hunk 13]
```diff
@@ -383,60 +324,57 @@ func (b *Builder) loadChallenge() error {
...
-			return fmt.Errorf("cannot create Nipst: %v", err)
+	}
+	pubEpoch := b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch)
+
+	hash, err := b.challenge.Hash()
+	if err != nil {
+		return fmt.Errorf("getting challenge hash failed: %v", err)
+	}
+	// ⏳ the following method waits for a PoET proof, which should take ~1 epoch
+	nipst, err := b.nipstBuilder.BuildNIPST(hash, b.layerClock.AwaitLayer((pubEpoch + 2).FirstLayer(b.layersPerEpoch)), b.stop)
```
Comment (antonlerner) [inline activation/activation.go]: why pass the channel and not the pubEpoch, setting timer inside BuildNIPST func? seems to me it is the nipstbuilders job to set it's own timeout

File: activation/activation.go [hunk 14]
```diff
@@ -383,60 +324,57 @@ func (b *Builder) loadChallenge() error {
...
+		return fmt.Errorf("failed to build nipst: %v", err)
+	}
+
+	b.log.With().Info("awaiting publication epoch",
+		log.Uint64("pub_epoch", uint64(pubEpoch)),
+		log.Uint64("pub_epoch_first_layer", uint64(pubEpoch.FirstLayer(b.layersPerEpoch))),
+		log.Uint64("current_layer", uint64(b.layerClock.GetCurrentLayer())),
+	)
+	select {
+	case <-b.layerClock.AwaitLayer(pubEpoch.FirstLayer(b.layersPerEpoch)):
```
Comment (antonlerner) [inline activation/activation.go]: have you tested what happens if the layer has passed already? will this not happen? also, why wait for the first layer of the publication epoch? shouldn't an ATX be pblished in `PubLayerIdx `

File: activation/activation.go [hunk 15]
```diff
@@ -456,114 +394,118 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
...
-	err = b.discardChallenge()
-	if err != nil {
-		log.Error("cannot discard Nipst challenge %s", err)
+	select {
+	case <-atxReceived:
+		b.log.Info("atx received in db")
+		events.Publish(events.AtxCreated{true, atx.ShortString(), uint64(b.currentEpoch())})
+	case <-b.layerClock.AwaitLayer((atx.TargetEpoch(b.layersPerEpoch) + 1).FirstLayer(b.layersPerEpoch)):
+		select {
+		case <-b.syncer.Await():
```
Comment (antonlerner) [inline activation/activation.go]: why wait to be synced here?

File: activation/nipst.go [hunk 16]
```diff
@@ -204,8 +205,15 @@ func (nb *NIPSTBuilder) BuildNIPST(challenge *types.Hash32) (*types.NIPST, error
...
-		proofRefChan := nb.poetDb.SubscribeToProofRef(nb.state.PoetServiceId, nb.state.PoetRound.Id)
-		poetProofRef := <-proofRefChan // TODO(noamnelke): handle timeout
+		var poetProofRef []byte
+		select {
+		case poetProofRef = <-nb.poetDb.SubscribeToProofRef(nb.state.PoetServiceId, nb.state.PoetRound.Id):
+		case <-timeout:
+			nb.poetDb.UnsubscribeFromProofRef(nb.state.PoetServiceId, nb.state.PoetRound.Id)
+			return nil, fmt.Errorf("timeout waiting for poet proof, atx target epoch ended")
+		case <-stop:
+			return nil, &StopRequestedError{}
```
Comment (antonlerner) [inline activation/nipst.go:215]: shouldnt unsubscribe here as well?

File: sync/syncer.go [hunk 17]
```diff
@@ -166,12 +167,28 @@ func (s *Syncer) ListenToGossip() bool {
...
 }
 
+func (s *Syncer) notifySubscribers(prevStatus, status Status) {
+	if (status == Done) == (prevStatus == Done) {
+		return
+	}
+	if status == Done {
+		close(s.awaitCh)
+	} else {
+		s.awaitCh = make(chan struct{})
```
Comment (antonlerner) [inline sync/syncer.go:184]: why create a channel here? this has nothing to do with notifying subscribers

File: activation/activation.go (hunk 10 above)
Comment (noamnelke) [reply activation/activation.go:188]: This was a lazy choice. This is what the previous implementation did.

I know that one class of errors, specifically PoET errors, will require special handling. There are three kinds of PoET errors: 
1. No response to registration requests - if the node is aware of another PoET _(no implemented)_ it should try another one.
2. PoET membership proof doesn't include our registration and we have a signed receipt _(not implemented)_ - blacklist PoET _(not implemented)_ and publish fraud proof _(not implemented)_.
3. Waiting for PoET proof timed out - blacklist PoET _(not implemented)_. Proving this happened is non-trivial, perhaps we can piggyback the hare, but not sure it's worth it.

As you can see none of this is possible right now (created issue for this just now #1673).

We could map out the other errors and see how we should handle each one, maybe the best handling for most of them really is to just retry immediately. I'll look at it now.

File: activation/activation.go (hunk 11 above)
Comment (noamnelke) [reply activation/activation.go:214]: I must admit I didn't think about this (this was the existing behaviour).

The risk if we keep the current behaviour is that the node registers to PoET with this challenge and then restarts at some point, essentially losing this epoch, since there's no ability to recover.

I guess it's better to fail fast. Good catch. Will change this.

File: activation/activation.go (hunk 12 above)
Comment (noamnelke) [reply activation/activation.go]: Feel free to suggest a better name.

What this method does is ensure that the challenge is "fresh" (the target epoch hasn't passed) and discards the challenge (and logs it) otherwise.

I could call it `discardChallengeIfStale`, wdyt?

File: activation/activation.go (hunk 13 above)
Comment (noamnelke) [reply activation/activation.go]: I disagree. The NISPT builder has no concept of time or layers, why involve it in this decision? It shouldn't be aware of the fact that an ATX is valid until the end of it's target epoch, or that the target epoch is one after the publication epoch. These are ATX builder concerns. It calls the NIPST builder and tells it "build me a NIPST before this time or return an error" which makes perfect sense to me.

I'm actually creating this same timeout in another place, so I think I'll extract it and call it `atxExpired` instead of timeout which sounds arbitrary.

File: activation/activation.go (hunk 14 above)
Comment (noamnelke) [reply activation/activation.go]: What you're suggesting would change the existing behaviour and isn't justified at the moment. The only reason we wait here is to have a full view of the previous epoch and there's no reason for a miner to wait any longer. Barak even went further and said there's no incentive to even wait this long, but I implemented it this way because it improves performance (everyone having the same view) and there's not too much to gain from publishing sooner.

Actually, thinking about it a little, in the current implementation `PubLayerIdx` is always the first layer of the epoch, so it would actually be exactly the same.

Not sure what you mean by "not happen". `AwaitLayer` returns a closed (non-blocking) channel if it's called for a past layer, so if that's what you mean - we won't block if the layer passed. This feature of `AwaitLayer` is tested in the clock:
https://github.com/spacemeshos/go-spacemesh/blob/5198d32ae48cb92bc5c8c493b8e193a42a9e8893/timesync/clock_test.go#L181-L192

File: activation/activation.go (hunk 15 above)
Comment (noamnelke) [reply activation/activation.go]: We're waiting for the ATX to appear in a block. We're waiting based on clock time, if the timeout expired, it could happen because nobody included our ATX, but it could also happen because we didn't receive the block yet because we're not synced. By waiting to be synced we ensure it's the first case.

There's currently a possibility that someone will include our ATX after the target epoch has ended causing our next ATX to be contextually invalid and us to be blacklisted due to it. #1655 should make this scenario impossible.

File: activation/nipst.go (hunk 16 above)
Comment (noamnelke) [reply activation/nipst.go:215]: Not sure what our policy here is - `stop` should happen only when the node shuts down, shouldn't it? What's the point in unsubscribing then? Wdyt?

File: sync/syncer.go (hunk 17 above)
Comment (noamnelke) [reply sync/syncer.go:184]: Currently subscribers just get whatever is in `s.awaitCh` which makes a simpler mechanism (I also create a channel in the constructor).

The alternative is to put a `nil` there and when someone subscribes do a nil-check and create the channel then, if needed, or return a closed channel. This would be more cumbersome.

Another advantage of the current implementation is that I don't need to lock the syncer for `Await()` calls, since it just returns whatever is in `s.awaitCh` (`notifySubscribers` is only called under lock anyway).

File: activation/activation.go (hunk 12 above)
Comment (noamnelke) [reply activation/activation.go]: I've renamed it to `discardChallengeIfStale` - I think it's really clearer this way

File: activation/activation.go (hunk 15 above)
Comment (noamnelke) [reply activation/activation.go]: I've added the comment 
```golang
// ensure we've seen all blocks before concluding that the ATX was lost
```

File: activation/activation.go [hunk 18]
```diff
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
 
 // Stop stops the atx builder.
 func (b *Builder) Stop() {
```
Comment (antonlerner) [inline activation/activation.go:142]: stop should unsubscribe from waiting on atx channel

File: activation/activation.go [hunk 19]
```diff
@@ -456,114 +395,118 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
...
+		b.log.Info("atx received in db")
+		events.Publish(events.AtxCreated{Created: true, Id: atx.ShortString(), Layer: uint64(b.currentEpoch())})
+	case <-b.layerClock.AwaitLayer((atx.TargetEpoch(b.layersPerEpoch) + 1).FirstLayer(b.layersPerEpoch)):
+		select {
+		case <-b.syncer.Await(): // ensure we've seen all blocks before concluding that the ATX was lost
+		case <-b.stop:
+			return &StopRequestedError{}
+		}
+		select {
+		case <-atxReceived:
```
Comment (antonlerner) [inline activation/activation.go:444]: this channel should be unsubscribed from if calling stop since no one will read the data sent on it after stop, is there a unit test for these edge cases?

File: activation/activation.go [hunk 20]
```diff
@@ -456,114 +395,118 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
...
+			b.log.Info("atx received in db (in the last moment)")
+			events.Publish(events.AtxCreated{Created: true, Id: atx.ShortString(), Layer: uint64(b.currentEpoch())})
+		default:
+			// TODO: Unsubscribe from ATX (add subscriber counter)
+			b.log.With().Error("target epoch has passed before atx was added to database",
+				log.AtxId(atx.ShortString()))
+			b.discardChallenge()
+			return fmt.Errorf("target epoch has passed")
+		}
+	case <-b.stop:
```
Comment (antonlerner) [inline activation/activation.go:455]: a lot of duplicatrs reading from this channel, can you refactir this  maybe?

File: activation/atxdb.go (hunk 8 above)
Comment (antonlerner) [reply activation/atxdb.go:502]: is it the same lock under which atxchannels are added?

File: activation/nipst.go (hunk 16 above)
Comment (antonlerner) [reply activation/nipst.go:215]: stop could happen if we call stop mining as well no?

File: activation/atxdb.go [hunk 21]
```diff
@@ -44,21 +44,47 @@ type ActivationDb struct {
...
+func init() {
+	close(closedChan)
+}
+
+func (db *ActivationDb) AwaitAtx(id types.AtxId) chan struct{} {
+	db.Lock()
+	defer db.Unlock()
+
+	if _, err := db.atxs.Get(getAtxHeaderKey(id)); err == nil {
+		return closedChan
```
Comment (antonlerner) [inline activation/atxdb.go]: can you do this to multiple subscribers? pass the same closed channel?

File: activation/activation.go (hunk 18 above)
Comment (noamnelke) [reply activation/activation.go:142]: 👍 done

File: activation/activation.go (hunk 19 above)
Comment (noamnelke) [reply activation/activation.go:444]: This is not a problem because nobody ever sends data on the channel, it's only ever closed, which is non-blocking.

However, to avoid a leak (although it's tiny), I've implemented unsubscribing.

File: activation/activation.go (hunk 20 above)
Comment (noamnelke) [reply activation/activation.go:455]: I've simplified it a little.

**Comment by noamnelke** [comment]
bors merge

**Comment by noamnelke** [comment]
bors merge
